dataset/string.cpp

```diff
@@ -73,17 +79,39 @@ std::string do_to_string(const D &dataset, const std::string &id,
 
   if (!dataset.coords().empty()) {
     s << "Coordinates:\n";
-    for (const auto &[dim, var] : dataset.coords())
-      s << format_variable(dim.name(), var, dims);
+    const auto map = dataset.coords();
+
+    // We need to cast unordered_map to vector so it can be sorted
+    std::vector<std::pair<std::string, VariableConstView>> elem;
+
+    // Elsewhere we just need a vector of Dataset attributes,
+    // here we need to call a function on these attributes.
+    for (const auto &[dim, var] : map) {
+      std::pair<std::string, VariableConstView> p = {dim.name(), var};
+      elem.push_back(p);
+    }
+    std::sort(elem.begin(), elem.end(), comp);
+    for (const auto &[name, var] : elem)
+      s << format_variable(name, var, dims);
   }
   if (!dataset.attrs().empty()) {
     s << "Attributes:\n";
-    for (const auto &[name, var] : dataset.attrs())
+
+    const auto map = dataset.attrs();
+    std::vector<std::pair<std::string, VariableConstView>> elems(map.begin(),
+                                                                 map.end());
+    std::sort(elems.begin(), elems.end(), comp);
+    for (const auto &[name, var] : elems)
       s << format_variable(name, var, dims);
   }
   if (!dataset.masks().empty()) {
     s << "Masks:\n";
-    for (const auto &[name, var] : dataset.masks())
+
+    const auto map = dataset.masks();
+    std::vector<std::pair<std::string, VariableConstView>> elems(map.begin(),
+                                                                 map.end());
+    std::sort(elems.begin(), elems.end(), comp);
+    for (const auto &[name, var] : elems)
```

Comment:
Looking that this pattern (almost) repeating 3 times: What we would like to write is
```cpp
for (const auto &[name, var] : sorted(dataset.masks())
```

You should be able of moving the duplicate lines to a `sorted` template to define above:
```cpp
template <class T>
auto sorted(const T &map) {
    std::vector<std::pair<std::string, VariableConstView>> elems(map.begin(), map.end());
    std::sort(elems.begin(), elems.end(), comp);
    return elems;
}
```

Of course this annoying does not work with `coords()`, but we can work around that, I think:
```cpp
template <class T>
auto sorted(const T &map) {
  std::vector<std::string> elems;
  std::transform(map.keys_begin(), map.keys_end(), [](const auto &key) { return to_string(key); });
  std::sort(elems.begin(), elems.end());
  return elems;
}
```
When calling this you need to make a minor change (or adapt the `sorted` function templae a bit more, whichever you prefer):
```cpp
    for (const auto &name : sorted(dataset.masks()))
      s << format_variable(name, dataset.masks[name], dims);
```

Reply:
Did something inbetween these two, since std::transform doesn't seem to handle our data types willingly.

Reply:
> Did something inbetween these two, since std::transform doesn't seem to handle our data types willingly.
That's because I messed up the syntax. ;) (I think `std::back_inserter` would have done the trick.